runtime/caml/skiplist.h

```diff
@@ -0,0 +1,102 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*             Xavier Leroy, projet Cambium, INRIA Paris                  */
+/*                                                                        */
+/*   Copyright 2020 Institut National de Recherche en Informatique et     */
+/*     en Automatique.                                                    */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+/* A dictionary data structure implemented as skip lists */
+
+/* Keys and associated data are natural-width integers (type [uintnat]).
+   Pointers can be used too, modulo conversion to [uintnat]. */
+
+#ifndef CAML_SKIPLIST_H
+#define CAML_SKIPLIST_H
+
+#include "config.h"
+
+#define NUM_LEVELS 17
+
+/* The head of a skip list */
+
+struct skiplist {
+  uintnat header1, header2;    /* dummy values for layout compatibility */
+  struct skipcell * forward[NUM_LEVELS]; /* forward chaining */
+  int level;                   /* max level used */
+};
+
+/* The cells of a skip list */
+
+struct skipcell {
+  uintnat key;
+  uintnat data;
+#if (__STDC_VERSION__ >= 199901L)
+  struct skipcell * forward[];  /* variable-length array */
+#else
+  struct skipcell * forward[1]; /* variable-length array */
+#endif
+};
+
+/* Size of struct skipcell, in bytes, without the forward array */
+#if (__STDC_VERSION__ >= 199901L)
+#define SIZEOF_SKIPCELL sizeof(struct skipcell)
+#else
+#define SIZEOF_SKIPCELL (sizeof(struct skipcell) - sizeof(struct skipcell *))
+#endif
+
+/* Initialize a skip list, statically */
+#define SKIPLIST_STATIC_INITIALIZER { 0, 0, {0, }, 0 }
+
+/* Initialize a skip list, dynamically */
+extern void caml_skiplist_init(struct skiplist * sk);
```

Comment:
This is a library, so it should include reasonably useful functions, even when not immediately used.  It is reasonable to have a dynamic initialization function for dynamically-allocated skip lists.

Reply:
As you prefer. Another option would be to have a const global variable initialized with what currently is `SKIPLIST_STATIC_INITIALIZER`, so that one can assign any skiplist (statically or dynamically allocated) from this const variable.

Reply:
Initializing a global variable from a const global variable is C++, but not C.  

To assign to a dynamically-allocated thing, we could use compound literals from C99 but I don't know if MSVC handles them.  So, let's do it the old-fashioned way, with an initialization function.

Reply:
Ah, you're right. But GCC accepts this without complaining with `-W -Wall -pedantic -std=c89`...

Reply:
Compound literals are MSVC 2013+

Reply:
This is the example I had in mind.  It is rejected by GCC (for all values of `-std`) but accepted by G++:
```
struct s { int x, y; };
const struct s s_init = { 1, 2 };
struct s s_dyn = s_init;
```

Reply:
This example compiles with my version of gcc (10.1.1). According to godbolt, it's accepted by gcc since version 8.

Reply:
I tried with GCC 7.5 (Ubuntu 18.04 LTS).  It's surprising that GCC would implement such an extension yet not complain if it's used in -pedantic mode.